Import BeautifulSoup and requests to extract html document from a url 

## All required imports

In [1]:
from bs4 import BeautifulSoup
import requests
import pandas as pd
import io #geocoder # import geocoder

import numpy as np # library to handle data in a vectorized manner

#import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

#import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
#!conda install -c conda-forge scikit-learn 
#from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: | 
The environment is inconsistent, please check the package plan carefully
The following packages are causing the inconsistency:

  - defaults/noarch::ibm-wsrt-py37main-main==custom=2155
  - conda-forge/linux-64::pytorch==1.8.0=cpu_py37ha70c682_1
  - defaults/noarch::ibm-wsrt-py37main-keep==0.0.0=2155
done

# All requested packages already installed.

Solving environment: \ 
The environment is inconsistent, please check the package plan carefully
The following packages are causing the inconsistency:

  - defaults/noarch::ibm-wsrt-py37main-main==custom=2155
  - conda-forge/linux-64::pytorch==1.8.0=cpu_py37ha70c682_1
  - defaults/noarch::ibm-wsrt-py37main-keep==0.0.0=2155
done

# All requested packages already installed.

Libraries imported.


## Get links

In [2]:
url="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv"
r=requests.get(url).content
pc=pd.read_csv(io.StringIO(r.decode('utf-8')))
pc.head(5)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Get BeautifulSoup object from url

In [3]:
req = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
soup = BeautifulSoup(req.content,'lxml')

Extract table containing postal code, borough and neighborhood

In [4]:
table=soup.find('table')

Generate a list to store information in. From the entire table, extract postal code, borough and neighborhood. Only extract information when boroughs are not 'Not assigned'. 
Since the table from the BeautifulSoup object is listed in rows we can loop through the rows to extract the information one by one and append the table_contents list by the cell objects.

In [5]:
table_contents=[]

for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)



Turn the list into a dataframe

In [6]:
# print(table_contents)
df=pd.DataFrame(table_contents)


Replace faulty borough information 

In [7]:
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

Display the shape of the dataframe

In [8]:
df.shape

(103, 3)

In [9]:
df.head(5)

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

In [10]:
df = df.merge(pc, how='left', left_on='PostalCode', right_on='Postal Code')

In [11]:
del df['Postal Code']

In [12]:
df.head(20)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A      Queen's Park   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   
15        M5C  Downtown Toronto   
16        M6C              York   
17        M9C         Etobicoke   
18        M1E       Scarborough   
19        M4E      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                       Ontario Provincial Government  43.662301 -79.389494  
5                                    Islington Avenue  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                     Don Mills North  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724  
12             Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
13                                    Don Mills South  43.725900 -79.340923  
14                                   Woodbine Heights  43.695344 -79.318389  
15                                     St. James Town  43.651494 -79.375418  
16                                 Humewood-Cedarvale  43.693781 -79.428191  
17  Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.643515 -79.577201  
18                  Guildwood, Morningside, West Hill  43.763573 -79.188711  
19                                        The Beaches  43.676357 -79.293031

Get geographical data for first mapping of neighborhoods

In [16]:
address = 'Toronto'

geolocator = Nominatim(user_agent="tr_exp")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


## Mapping of Toronto Neighborhoods

In [17]:
#map of toronto
map_tr = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_tr)  
    
map_tr

### Only include boroughs that have york in name

In [18]:
tr_york = df[df['Borough'].str.contains("York")].reset_index(drop=True)
tr_york.head(30)

PostalCode                 Borough  \
0         M3A              North York   
1         M4A              North York   
2         M6A              North York   
3         M3B              North York   
4         M4B               East York   
5         M6B              North York   
6         M3C              North York   
7         M4C               East York   
8         M6C                    York   
9         M6E                    York   
10        M4G               East York   
11        M2H              North York   
12        M3H              North York   
13        M4H               East York   
14        M2J              North York   
15        M3J              North York   
16        M4J  East York/East Toronto   
17        M2K              North York   
18        M3K              North York   
19        M2L              North York   
20        M3L              North York   
21        M6L              North York   
22        M9L              North York   
23        M2M              North York   
24        M3M              North York   
25        M5M              North York   
26        M6M                    York   
27        M9M              North York   
28        M2N              North York   
29        M3N              North York   

                                        Neighborhood   Latitude  Longitude  
0                                          Parkwoods  43.753259 -79.329656  
1                                   Victoria Village  43.725882 -79.315572  
2                   Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
3                                    Don Mills North  43.745906 -79.352188  
4                    Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
5                                          Glencairn  43.709577 -79.445073  
6                                    Don Mills South  43.725900 -79.340923  
7                                   Woodbine Heights  43.695344 -79.318389  
8                                 Humewood-Cedarvale  43.693781 -79.428191  
9                                Caledonia-Fairbanks  43.689026 -79.453512  
10                                           Leaside  43.709060 -79.363452  
11                                 Hillcrest Village  43.803762 -79.363452  
12   Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259  
13                                  Thorncliffe Park  43.705369 -79.349372  
14                      Fairview, Henry Farm, Oriole  43.778517 -79.346556  
15                   Northwood Park, York University  43.767980 -79.487262  
16                                The Danforth  East  43.685347 -79.338106  
17                                   Bayview Village  43.786947 -79.385975  
18                                    Downsview East  43.737473 -79.464763  
19                          York Mills, Silver Hills  43.757490 -79.374714  
20                                    Downsview West  43.739015 -79.506944  
21          North Park, Maple Leaf Park, Upwood Park  43.713756 -79.490074  
22                                     Humber Summit  43.756303 -79.565963  
23                           Willowdale, Newtonbrook  43.789053 -79.408493  
24                                 Downsview Central  43.728496 -79.495697  
25                 Bedford Park, Lawrence Manor East  43.733283 -79.419750  
26  Del Ray, Mount Dennis, Keelsdale and Silverthorn  43.691116 -79.476013  
27                                  Humberlea, Emery  43.724766 -79.532242  
28                                  Willowdale South  43.770120 -79.408493  
29                               Downsview Northwest  43.761631 -79.520999

## Map again, this time only boroughs with york in name

In [19]:
# create map of the yorks using latitude and longitude values
map_yorks = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(tr_york['Latitude'], tr_york['Longitude'], tr_york['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_yorks)  
    
map_yorks

## Counting of unique neighborhoods in boroughs with York in the name

As can be seen, the borough of East York / East Toronto only has one neighborhood, while the borough North York has 24

In [20]:
df_test = tr_york.groupby('Borough')['Neighborhood'].nunique()

print (df_test)

Borough
East York                  4
East York/East Toronto     1
North York                24
York                       5
Name: Neighborhood, dtype: int64


### Boroughs are numbered for mapping preparation and then mapped based on their numbering

In [21]:
tr_york_br = tr_york
# create a list of our conditions
conditions = [
    (tr_york_br['Borough'] == "York"),
    (tr_york_br['Borough'] == "East York"),
    (tr_york_br['Borough'] == "North York"),
    (tr_york_br['Borough'] == "East York/East Toronto")
    ]

# create a list of the values we want to assign for each condition
values = [4, 3, 2, 1]

# create a new column and use np.select to assign values to it using our lists as arguments
tr_york_br['br_n'] = np.select(conditions, values)

# display updated DataFrame
tr_york_br.head()


PostalCode     Borough                      Neighborhood   Latitude  \
0        M3A  North York                         Parkwoods  43.753259   
1        M4A  North York                  Victoria Village  43.725882   
2        M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
3        M3B  North York                   Don Mills North  43.745906   
4        M4B   East York   Parkview Hill, Woodbine Gardens  43.706397   

   Longitude  br_n  
0 -79.329656     2  
1 -79.315572     2  
2 -79.464763     2  
3 -79.352188     2  
4 -79.309937     3

In [22]:
n_br = 4

# create map
map_bor = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the boroughs
x = np.arange(n_br)
ys = [i + x + (i*x)**2 for i in range(n_br)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, br in zip(tr_york_br['Latitude'], tr_york_br['Longitude'], tr_york_br['Neighborhood'], tr_york_br['br_n']):
    label = folium.Popup(str(poi) + ' Borough ' + str(br), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[br-1],
        fill=True,
        fill_color=rainbow[br-1],
        fill_opacity=0.7).add_to(map_bor)
       
map_bor

As can be seen, East York / East Toronto lies in the South East close to the river (purple dot)

In [23]:
CLIENT_ID = 'QW0D2R4LUIW15O4XC0SSZ1DOCBDKSQGU0ZM24Y25JUGYOPDB' # your Foursquare ID
CLIENT_SECRET = '3XE0G1EBHQQ0CZ40CPWKWW2ZIVTA2ACXAUDATE05UN2DZUWE' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: QW0D2R4LUIW15O4XC0SSZ1DOCBDKSQGU0ZM24Y25JUGYOPDB
CLIENT_SECRET:3XE0G1EBHQQ0CZ40CPWKWW2ZIVTA2ACXAUDATE05UN2DZUWE


### Neighborhood data is determined

In [24]:
neighborhood_latitude = tr_york_br.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = tr_york_br.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = tr_york_br.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))


Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


### 2000 metres radius around the neighborhood location is chosen, url for foursquare request is generated

In [25]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 2000 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=QW0D2R4LUIW15O4XC0SSZ1DOCBDKSQGU0ZM24Y25JUGYOPDB&client_secret=3XE0G1EBHQQ0CZ40CPWKWW2ZIVTA2ACXAUDATE05UN2DZUWE&v=20180605&ll=43.7532586,-79.3296565&radius=2000&limit=100'

In [26]:
results_york = requests.get(url).json()
results_york

{'meta': {'code': 200, 'requestId': '609068090e6a173654a77b06'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 89,
  'suggestedBounds': {'ne': {'lat': 43.77125861800002,
    'lng': -79.30478345939711},
   'sw': {'lat': 43.735258581999986, 'lng': -79.35452954060288}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b8991cbf964a520814232e3',
       'name': "Allwyn's Bakery",
       'location': {'address': '81 Underhill drive',
        'lat': 43.75984035203157,
        'lng': -79.32471879917513,
        'labeledLatLngs': [{'label': 'display'

## Now that the locations of interest are found, the data is prepared so that the venues can be analyzed

In [27]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [28]:
venues = results_york['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]
nearby_venues.head()


/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  app.launch_new_instance()


venue.name  \
0              Allwyn's Bakery   
1                  Tim Hortons   
2  Donalda Golf & Country Club   
3           Graydon Hall Manor   
4                 Island Foods   

                                    venue.categories  venue.location.lat  \
0  [{'id': '4bf58dd8d48988d144941735', 'name': 'C...           43.759840   
1  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...           43.760668   
2  [{'id': '4bf58dd8d48988d1e6941735', 'name': 'G...           43.752816   
3  [{'id': '4bf58dd8d48988d171941735', 'name': 'E...           43.763923   
4  [{'id': '4bf58dd8d48988d144941735', 'name': 'C...           43.745866   

   venue.location.lng  
0          -79.324719  
1          -79.326368  
2          -79.342741  
3          -79.342961  
4          -79.346035

In [29]:
test_lst = nearby_venues['venue.categories']
testset = test_lst[0]
testset
pd_test = pd.DataFrame(testset)
pd_test['name']
pd_test

id                  name             pluralName  \
0  4bf58dd8d48988d144941735  Caribbean Restaurant  Caribbean Restaurants   

   shortName                                               icon  primary  
0  Caribbean  {'prefix': 'https://ss3.4sqi.net/img/categorie...     True

## The name of the venue category is extracted through the function here under

In [30]:
def get_name(dframe):
    for index, row in dframe.iterrows(): 
        lst = dframe['venue.categories']
        name_df = pd.DataFrame(lst[index])
        name = name_df['name'][0]
        dframe['venue.categories'][index] = name
    return dframe
    

In [31]:
nearby_venues['venue.categories']
lst_names = get_name(nearby_venues)
nearby_venues_tr = lst_names
nearby_venues_tr.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


venue.name      venue.categories  venue.location.lat  \
0              Allwyn's Bakery  Caribbean Restaurant           43.759840   
1                  Tim Hortons                  Café           43.760668   
2  Donalda Golf & Country Club           Golf Course           43.752816   
3           Graydon Hall Manor           Event Space           43.763923   
4                 Island Foods  Caribbean Restaurant           43.745866   

   venue.location.lng  
0          -79.324719  
1          -79.326368  
2          -79.342741  
3          -79.342961  
4          -79.346035

The venue categories are found and can be counted

In [33]:
print('{} venues were returned by Foursquare.'.format(nearby_venues_tr.shape[0]))

89 venues were returned by Foursquare.


### As in the New York example, nearby venues are determined

In [34]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results_york = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results_york])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [35]:
tr_york_venues = getNearbyVenues(names=tr_york_br['Neighborhood'],
                                   latitudes=tr_york_br['Latitude'],
                                   longitudes=tr_york_br['Longitude']
                                  )

Parkwoods
Victoria Village
Lawrence Manor, Lawrence Heights
Don Mills North
Parkview Hill, Woodbine Gardens
Glencairn
Don Mills South
Woodbine Heights
Humewood-Cedarvale
Caledonia-Fairbanks
Leaside
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Bayview Village
Downsview East
York Mills, Silver Hills
Downsview West
North Park, Maple Leaf Park, Upwood Park
Humber Summit
Willowdale, Newtonbrook
Downsview Central
Bedford Park, Lawrence Manor East
Del Ray, Mount Dennis, Keelsdale and Silverthorn
Humberlea, Emery
Willowdale South
Downsview Northwest
Runnymede, The Junction North
Weston
York Mills West
Willowdale West


In [36]:
print(tr_york_venues.shape)
tr_york_venues.head()

(337, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3               Portugril       43.725819       -79.312785   
4             Tim Hortons       43.725517       -79.313103   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3  Portuguese Restaurant  
4            Coffee Shop

In [37]:
tr_york_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                              
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    26   
Caledonia-Fairbanks                                                   4   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                      4   
Don Mills North                                                       5   
Don Mills South                                                      19   
Downsview Central                                                     4   
Downsview East                                                        3   
Downsview Northwest                                                   4   
Downsview West                                                        6   
Fairview, Henry Farm, Oriole                                         68   
Glencairn                                                             5   
Hillcrest Village                                                     6   
Humber Summit                                                         1   
Humberlea, Emery                                                      2   
Humewood-Cedarvale                                                    4   
Lawrence Manor, Lawrence Heights                                     14   
Leaside                                                              31   
North Park, Maple Leaf Park, Upwood Park                              5   
Northwood Park, York University                                       7   
Parkview Hill, Woodbine Gardens                                      11   
Parkwoods                                                             2   
Runnymede, The Junction North                                         3   
The Danforth  East                                                    4   
Thorncliffe Park                                                     21   
Victoria Village                                                      4   
Weston                                                                1   
Willowdale South                                                     34   
Willowdale West                                                       5   
Willowdale, Newtonbrook                                               1   
Woodbine Heights                                                      5   
York Mills West                                                       3   

                                                  Neighborhood Longitude  \
Neighborhood                                                               
Bathurst Manor, Wilson Heights, Downsview North                       21   
Bayview Village                                                        4   
Bedford Park, Lawrence Manor East                                     26   
Caledonia-Fairbanks                                                    4   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                       4   
Don Mills North                                                        5   
Don Mills South                                                       19   
Downsview Central                                                      4   
Downsview East                                                         3   
Downsview Northwest                                                    4   
Downsview West                                                         6   
Fairview, Henry Farm, Oriole                                          68   
Glencairn                                                              5   
Hillcrest Village                                                      6   
Humber Summit                                                          1   
Humberlea, Emery                                                       2   
Humewood-Cedarvale                                      

## The counted venues by neighborhood can be extracted from the table above. As can be seen, Fairview, Henry Farm and Oriole have the most venues (68), followed by Willowdale South (34) and Leaside (31). Only 3 neighborhoods have only 1 venue.

In [38]:
print('There are {} uniques categories.'.format(len(tr_york_venues['Venue Category'].unique())))

There are 118 uniques categories.


After it is found that there are 118 categories, they are attributed to the neighborhoods as shown below

In [39]:
# one hot encoding
tr_york_onehot = pd.get_dummies(tr_york_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
tr_york_onehot['Neighborhood'] = tr_york_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [tr_york_onehot.columns[-1]] + list(tr_york_onehot.columns[:-1])
tr_york_onehot = tr_york_onehot[fixed_columns]

tr_york_onehot.head()

Neighborhood  Accessories Store  Airport  American Restaurant  \
0         Parkwoods                  0        0                    0   
1         Parkwoods                  0        0                    0   
2  Victoria Village                  0        0                    0   
3  Victoria Village                  0        0                    0   
4  Victoria Village                  0        0                    0   

   Art Gallery  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0            0                    0                 0                   0   
1            0                    0                 0                   0   
2            0                    0                 0                   0   
3            0                    0                 0                   0   
4            0                    0                 0                   0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Basketball Court  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    0               0                 0   
3           0       0     0    0               0                 0   
4           0       0     0    0               0                 0   

   Beer Store  Bike Shop  Boutique  Breakfast Spot  Brewery  Bridal Shop  \
0           0          0         0               0        0            0   
1           0          0         0               0        0            0   
2           0          0         0               0        0            0   
3           0          0         0               0        0            0   
4           0          0         0               0        0            0   

   Bubble Tea Shop  Burger Joint  Burrito Place  Business Service  Butcher  \
0                0             0              0                 0        0   
1                0             0              0                 0        0   
2                0             0              0                 0        0   
3                0             0              0                 0        0   
4                0             0              0                 0        0   

   Café  Caribbean Restaurant  Chinese Restaurant  Chocolate Shop  \
0     0                     0                   0               0   
1     0                     0                   0               0   
2     0                     0                   0               0   
3     0                     0                   0               0   
4     0                     0                   0               0   

   Clothing Store  Coffee Shop  Comfort Food Restaurant  \
0               0            0                        0   
1               0            0                        0   
2               0            0                        0   
3               0            0                        0   
4               0            1                        0   

   Construction & Landscaping  Convenience Store  Cosmetics Shop  \
0                           0                  0               0   
1                           0                  0               0   
2                           0                  0               0   
3                           0                  0               0   
4                           0                  0               0   

   Cupcake Shop  Curling Ice  Deli / Bodega  Department Store  Dessert Shop  \
0             0            0              0                 0             0   
1             0            0              0                 0             0   
2             0            0              0                 0             0   
3             0            0              0                 0             0   
4             0            0              0                 0             0   

   Dim Sum Restaurant  Diner  Discount Store  Distribution Center  Dog Run  \
0                   0      0    

In [40]:
tr_york_grouped = tr_york_onehot.groupby('Neighborhood').mean().reset_index()
tr_york_grouped

Neighborhood  Accessories Store  \
0    Bathurst Manor, Wilson Heights, Downsview North           0.000000   
1                                    Bayview Village           0.000000   
2                  Bedford Park, Lawrence Manor East           0.000000   
3                                Caledonia-Fairbanks           0.000000   
4   Del Ray, Mount Dennis, Keelsdale and Silverthorn           0.000000   
5                                    Don Mills North           0.000000   
6                                    Don Mills South           0.000000   
7                                  Downsview Central           0.000000   
8                                     Downsview East           0.000000   
9                                Downsview Northwest           0.000000   
10                                    Downsview West           0.000000   
11                      Fairview, Henry Farm, Oriole           0.000000   
12                                         Glencairn           0.000000   
13                                 Hillcrest Village           0.000000   
14                                     Humber Summit           0.000000   
15                                  Humberlea, Emery           0.000000   
16                                Humewood-Cedarvale           0.000000   
17                  Lawrence Manor, Lawrence Heights           0.142857   
18                                           Leaside           0.000000   
19          North Park, Maple Leaf Park, Upwood Park           0.000000   
20                   Northwood Park, York University           0.000000   
21                   Parkview Hill, Woodbine Gardens           0.000000   
22                                         Parkwoods           0.000000   
23                     Runnymede, The Junction North           0.000000   
24                                The Danforth  East           0.000000   
25                                  Thorncliffe Park           0.000000   
26                                  Victoria Village           0.000000   
27                                            Weston           0.000000   
28                                  Willowdale South           0.000000   
29                                   Willowdale West           0.000000   
30                           Willowdale, Newtonbrook           0.000000   
31                                  Woodbine Heights           0.000000   
32                                   York Mills West           0.000000   

     Airport  American Restaurant  Art Gallery  Arts & Crafts Store  \
0   0.000000             0.000000     0.000000             0.000000   
1   0.000000             0.000000     0.000000             0.000000   
2   0.000000             0.038462     0.000000             0.000000   
3   0.000000             0.000000     0.000000             0.000000   
4   0.000000             0.000000     0.000000             0.000000   
5   0.000000             0.000000     0.000000             0.000000   
6   0.000000             0.000000     0.052632             0.000000   
7   0.000000             0.000000     0.000000             0.000000   
8   0.333333             0.000000     0.000000             0.000000   
9   0.000000             0.000000     0.000000             0.000000   
10  0.000000             0.000000     0.000000             0.000000   
11  0.000000             0.014706     0.000000             0.000000   
12  0.000000             0.000000     0.000000             0.000000   
13  0.000000             0.000000     0.000000             0.000000   
14  0.000000             0.000000     0.000000             0.000000   
15  0.000000             0.000000     0.000000             0.000000   
16  0.000000             0.000000     0.000000             0.000000   
17  0.000000             0.000000     0.000000             0.000000   
18  0.000000             0.000000     0.000000             0.000000   
19  0.000000             0.000000     0.000000             0.000000  

### Above, you can find the full list of neighborhoods and their corresponding venue categories

In [41]:
tr_york_grouped.shape

(33, 119)

The formula below is used to determine the most common venues by neighborhood

In [42]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

## Below, the most common venue categories by neighborhood are listed

In [43]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = tr_york_grouped['Neighborhood']

for ind in np.arange(tr_york_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(tr_york_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0   Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
1                                   Bayview Village    Chinese Restaurant   
2                 Bedford Park, Lawrence Manor East            Restaurant   
3                               Caledonia-Fairbanks                  Park   
4  Del Ray, Mount Dennis, Keelsdale and Silverthorn     Convenience Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Bank              Pharmacy           Supermarket   
1                  Café                  Bank   Japanese Restaurant   
2    Italian Restaurant           Coffee Shop        Sandwich Place   
3         Women's Store                  Pool             Juice Bar   
4           Coffee Shop        Discount Store        Sandwich Place   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Middle Eastern Restaurant                 Diner     Health Food Store   
1          Accessories Store                Lounge         Moving Target   
2                Pizza Place               Butcher                  Café   
3              Movie Theater     Mobile Phone Shop    Miscellaneous Shop   
4                     Lounge         Moving Target         Movie Theater   

       8th Most Common Venue    9th Most Common Venue  \
0                Pizza Place               Restaurant   
1              Movie Theater        Mobile Phone Shop   
2                        Pub  Comfort Food Restaurant   
3  Middle Eastern Restaurant       Mexican Restaurant   
4          Mobile Phone Shop       Miscellaneous Shop   

      10th Most Common Venue  
0             Sandwich Place  
1         Miscellaneous Shop  
2           Greek Restaurant  
3              Metro Station  
4  Middle Eastern Restaurant

In [44]:
!pip uninstall -y numpy
!pip uninstall -y setuptools


Found existing installation: numpy 1.20.2
Uninstalling numpy-1.20.2:
  Successfully uninstalled numpy-1.20.2
Found existing installation: setuptools 56.0.0
Uninstalling setuptools-56.0.0:
  Successfully uninstalled setuptools-56.0.0


In [45]:
!pip install setuptools
!pip install numpy

  Using cached setuptools-56.0.0-py3-none-any.whl (784 kB)
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorboard 1.14.0 requires numpy>=1.12.0, which is not installed.
ibmdbpy 0.1.6 requires numpy, which is not installed.
altair 4.1.0 requires numpy, which is not installed.
  Using cached numpy-1.20.2-cp37-cp37m-manylinux2010_x86_64.whl (15.3 MB)


In [46]:
# import k-means from clustering stage
!pip install -U scikit-learn

# Clustering

In [47]:
from sklearn.cluster import KMeans

## 4 Clusters are used to correlate results to the 4 boroughs and to notice differences

In [48]:
# set number of clusters
kclusters = 4

tr_york_grouped_clustering = tr_york_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(tr_york_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 0, 1, 1, 1, 1, 1, 1], dtype=int32)

In [49]:
kmeans

KMeans(n_clusters=4, random_state=0)

In [50]:
neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0   Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
1                                   Bayview Village    Chinese Restaurant   
2                 Bedford Park, Lawrence Manor East            Restaurant   
3                               Caledonia-Fairbanks                  Park   
4  Del Ray, Mount Dennis, Keelsdale and Silverthorn     Convenience Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Bank              Pharmacy           Supermarket   
1                  Café                  Bank   Japanese Restaurant   
2    Italian Restaurant           Coffee Shop        Sandwich Place   
3         Women's Store                  Pool             Juice Bar   
4           Coffee Shop        Discount Store        Sandwich Place   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Middle Eastern Restaurant                 Diner     Health Food Store   
1          Accessories Store                Lounge         Moving Target   
2                Pizza Place               Butcher                  Café   
3              Movie Theater     Mobile Phone Shop    Miscellaneous Shop   
4                     Lounge         Moving Target         Movie Theater   

       8th Most Common Venue    9th Most Common Venue  \
0                Pizza Place               Restaurant   
1              Movie Theater        Mobile Phone Shop   
2                        Pub  Comfort Food Restaurant   
3  Middle Eastern Restaurant       Mexican Restaurant   
4          Mobile Phone Shop       Miscellaneous Shop   

      10th Most Common Venue  
0             Sandwich Place  
1         Miscellaneous Shop  
2           Greek Restaurant  
3              Metro Station  
4  Middle Eastern Restaurant

## Cluster labels are added below

In [51]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_,allow_duplicates=True)

neighborhoods_venues_sorted.head()

Cluster Labels                                      Neighborhood  \
0               1   Bathurst Manor, Wilson Heights, Downsview North   
1               1                                   Bayview Village   
2               1                 Bedford Park, Lawrence Manor East   
3               0                               Caledonia-Fairbanks   
4               1  Del Ray, Mount Dennis, Keelsdale and Silverthorn   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop                  Bank              Pharmacy   
1    Chinese Restaurant                  Café                  Bank   
2            Restaurant    Italian Restaurant           Coffee Shop   
3                  Park         Women's Store                  Pool   
4     Convenience Store           Coffee Shop        Discount Store   

  4th Most Common Venue      5th Most Common Venue 6th Most Common Venue  \
0           Supermarket  Middle Eastern Restaurant                 Diner   
1   Japanese Restaurant          Accessories Store                Lounge   
2        Sandwich Place                Pizza Place               Butcher   
3             Juice Bar              Movie Theater     Mobile Phone Shop   
4        Sandwich Place                     Lounge         Moving Target   

  7th Most Common Venue      8th Most Common Venue    9th Most Common Venue  \
0     Health Food Store                Pizza Place               Restaurant   
1         Moving Target              Movie Theater        Mobile Phone Shop   
2                  Café                        Pub  Comfort Food Restaurant   
3    Miscellaneous Shop  Middle Eastern Restaurant       Mexican Restaurant   
4         Movie Theater          Mobile Phone Shop       Miscellaneous Shop   

      10th Most Common Venue  
0             Sandwich Place  
1         Miscellaneous Shop  
2           Greek Restaurant  
3              Metro Station  
4  Middle Eastern Restaurant

In [52]:
tr_york_merged = tr_york_br

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
tr_york_merged = tr_york_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

tr_york_merged.head() # check the last columns!

PostalCode     Borough                      Neighborhood   Latitude  \
0        M3A  North York                         Parkwoods  43.753259   
1        M4A  North York                  Victoria Village  43.725882   
2        M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
3        M3B  North York                   Don Mills North  43.745906   
4        M4B   East York   Parkview Hill, Woodbine Gardens  43.706397   

   Longitude  br_n  Cluster Labels 1st Most Common Venue  \
0 -79.329656     2             0.0     Food & Drink Shop   
1 -79.315572     2             1.0           Coffee Shop   
2 -79.464763     2             1.0        Clothing Store   
3 -79.352188     2             1.0  Caribbean Restaurant   
4 -79.309937     3             1.0           Pizza Place   

  2nd Most Common Venue   3rd Most Common Venue  4th Most Common Venue  \
0                  Park       Accessories Store      Korean Restaurant   
1           Pizza Place            Hockey Arena  Portuguese Restaurant   
2     Accessories Store  Furniture / Home Store            Coffee Shop   
3                   Gym      Athletics & Sports    Japanese Restaurant   
4                  Bank                Pharmacy   Gym / Fitness Center   

  5th Most Common Venue      6th Most Common Venue  7th Most Common Venue  \
0         Movie Theater          Mobile Phone Shop     Miscellaneous Shop   
1          Liquor Store              Movie Theater      Mobile Phone Shop   
2              Boutique         Miscellaneous Shop  Vietnamese Restaurant   
3                  Café  Middle Eastern Restaurant         Massage Studio   
4           Flea Market               Intersection         Breakfast Spot   

       8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0  Middle Eastern Restaurant         Mexican Restaurant          Metro Station  
1         Miscellaneous Shop  Middle Eastern Restaurant     Mexican Restaurant  
2              Women's Store          Mobile Phone Shop          Movie Theater  
3   Mediterranean Restaurant              Metro Station     Mexican Restaurant  
4                  Pet Store         Athletics & Sports              Gastropub

In [53]:
tr_york_merged['Cluster Labels'].head()

0    0.0
1    1.0
2    1.0
3    1.0
4    1.0
Name: Cluster Labels, dtype: float64

### Cluster labels are converted to integers below to allow for formulas further down to run

In [54]:
tr_york_merged.fillna({'Cluster Labels':0.0}, inplace=True)
#tr_york_merged.astype({'Cluster Labels': int})
tr_york_merged = tr_york_merged.astype({'Cluster Labels': int})
tr_york_merged

PostalCode                 Borough  \
0         M3A              North York   
1         M4A              North York   
2         M6A              North York   
3         M3B              North York   
4         M4B               East York   
5         M6B              North York   
6         M3C              North York   
7         M4C               East York   
8         M6C                    York   
9         M6E                    York   
10        M4G               East York   
11        M2H              North York   
12        M3H              North York   
13        M4H               East York   
14        M2J              North York   
15        M3J              North York   
16        M4J  East York/East Toronto   
17        M2K              North York   
18        M3K              North York   
19        M2L              North York   
20        M3L              North York   
21        M6L              North York   
22        M9L              North York   
23        M2M              North York   
24        M3M              North York   
25        M5M              North York   
26        M6M                    York   
27        M9M              North York   
28        M2N              North York   
29        M3N              North York   
30        M6N                    York   
31        M9N                    York   
32        M2P              North York   
33        M2R              North York   

                                        Neighborhood   Latitude  Longitude  \
0                                          Parkwoods  43.753259 -79.329656   
1                                   Victoria Village  43.725882 -79.315572   
2                   Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
3                                    Don Mills North  43.745906 -79.352188   
4                    Parkview Hill, Woodbine Gardens  43.706397 -79.309937   
5                                          Glencairn  43.709577 -79.445073   
6                                    Don Mills South  43.725900 -79.340923   
7                                   Woodbine Heights  43.695344 -79.318389   
8                                 Humewood-Cedarvale  43.693781 -79.428191   
9                                Caledonia-Fairbanks  43.689026 -79.453512   
10                                           Leaside  43.709060 -79.363452   
11                                 Hillcrest Village  43.803762 -79.363452   
12   Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259   
13                                  Thorncliffe Park  43.705369 -79.349372   
14                      Fairview, Henry Farm, Oriole  43.778517 -79.346556   
15                   Northwood Park, York University  43.767980 -79.487262   
16                                The Danforth  East  43.685347 -79.338106   
17                                   Bayview Village  43.786947 -79.385975   
18                                    Downsview East  43.737473 -79.464763   
19                          York Mills, Silver Hills  43.757490 -79.374714   
20                                    Downsview West  43.739015 -79.506944   
21          North Park, Maple Leaf Park, Upwood Park  43.713756 -79.490074   
22                                     Humber Summit  43.756303 -79.565963   
23                           Willowdale, Newtonbrook  43.789053 -79.408493   
24                                 Downsview Central  43.728496 -79.495697   
25                 Bedford Park, Lawrence Manor East  43.733283 -79.419750   
26  Del Ray, Mount Dennis, Keelsdale and Silverthorn  43.691116 -79.476013   
27                                  Humberlea, Emery  43.724766 -79.532242   
28                                  Willowdale South  43.770120 -79.408493   
29                               Downsview Northwest  43.761631 -79.520999   
30                     Runnymede, The Junction North  43.673185 -79.487262   
31                                            Weston  43.706876 -79.518

### Clusters are mapped below

In [55]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tr_york_merged['Latitude'], tr_york_merged['Longitude'], tr_york_merged['Neighborhood'], tr_york_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### It can be seen that the clusters do not correspond to the boroughs as shown earlier. Two clusters only have one entry. The entry for East York / East Toronto (it's own borough) is now part of cluster 0. Cluster 1 has by far the most entries. Cluster 2 (cyan) and 3 (light green, top left, difficult to make out) only have one entry each. In the following, the clusters are explored.

In [56]:
tr_york_merged.loc[tr_york_merged['Cluster Labels'] == 0, tr_york_merged.columns[[1] + list(range(5, tr_york_merged.shape[1]))]]

Borough  br_n  Cluster Labels 1st Most Common Venue  \
0               North York     2               0     Food & Drink Shop   
9                     York     4               0                  Park   
16  East York/East Toronto     1               0                  Park   
19              North York     2               0                   NaN   
23              North York     2               0                  Park   
31                    York     4               0                  Park   
32              North York     2               0     Convenience Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Park     Accessories Store     Korean Restaurant   
9          Women's Store                  Pool             Juice Bar   
16     Convenience Store          Intersection     Accessories Store   
19                   NaN                   NaN                   NaN   
23     Accessories Store             Juice Bar         Movie Theater   
31     Accessories Store             Juice Bar         Movie Theater   
32                   Bar                  Park     Korean Restaurant   

   5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0          Movie Theater     Mobile Phone Shop         Miscellaneous Shop   
9          Movie Theater     Mobile Phone Shop         Miscellaneous Shop   
16          Liquor Store         Movie Theater          Mobile Phone Shop   
19                   NaN                   NaN                        NaN   
23     Mobile Phone Shop    Miscellaneous Shop  Middle Eastern Restaurant   
31     Mobile Phone Shop    Miscellaneous Shop  Middle Eastern Restaurant   
32         Moving Target         Movie Theater          Mobile Phone Shop   

        8th Most Common Venue      9th Most Common Venue  \
0   Middle Eastern Restaurant         Mexican Restaurant   
9   Middle Eastern Restaurant         Mexican Restaurant   
16         Miscellaneous Shop  Middle Eastern Restaurant   
19                        NaN                        NaN   
23         Mexican Restaurant              Metro Station   
31         Mexican Restaurant              Metro Station   
32         Miscellaneous Shop  Middle Eastern Restaurant   

      10th Most Common Venue  
0              Metro Station  
9              Metro Station  
16        Mexican Restaurant  
19                       NaN  
23  Mediterranean Restaurant  
31  Mediterranean Restaurant  
32        Mexican Restaurant

Cluster 0 is primarily made up of parks. This differentiates it from the other clusters and explains why the venues are spread across the city.

In [57]:
tr_york_merged.loc[tr_york_merged['Cluster Labels'] == 1, tr_york_merged.columns[[1] + list(range(5, tr_york_merged.shape[1]))]]

Borough  br_n  Cluster Labels       1st Most Common Venue  \
1   North York     2               1                 Coffee Shop   
2   North York     2               1              Clothing Store   
3   North York     2               1        Caribbean Restaurant   
4    East York     3               1                 Pizza Place   
5   North York     2               1                      Bakery   
6   North York     2               1                 Coffee Shop   
7    East York     3               1                Intersection   
8         York     4               1               Moving Target   
10   East York     3               1                 Coffee Shop   
11  North York     2               1        Fast Food Restaurant   
12  North York     2               1                 Coffee Shop   
13   East York     3               1           Indian Restaurant   
14  North York     2               1              Clothing Store   
15  North York     2               1        Caribbean Restaurant   
17  North York     2               1          Chinese Restaurant   
18  North York     2               1                        Park   
20  North York     2               1               Grocery Store   
21  North York     2               1  Construction & Landscaping   
24  North York     2               1                Home Service   
25  North York     2               1                  Restaurant   
26        York     4               1           Convenience Store   
28  North York     2               1            Ramen Restaurant   
29  North York     2               1               Grocery Store   
30        York     4               1               Grocery Store   
33  North York     2               1                 Coffee Shop   

       2nd Most Common Venue   3rd Most Common Venue  4th Most Common Venue  \
1                Pizza Place            Hockey Arena  Portuguese Restaurant   
2          Accessories Store  Furniture / Home Store            Coffee Shop   
3                        Gym      Athletics & Sports    Japanese Restaurant   
4                       Bank                Pharmacy   Gym / Fitness Center   
5                Pizza Place                    Park    Japanese Restaurant   
6                 Restaurant                     Gym     Italian Restaurant   
7                       Park             Curling Ice             Beer Store   
8               Hockey Arena                   Trail                  Field   
10    Furniture / Home Store            Burger Joint                   Bank   
11  Mediterranean Restaurant             Golf Course                Dog Run   
12                      Bank                Pharmacy            Supermarket   
13            Sandwich Place             Yoga Studio           Burger Joint   
14               Coffee Shop    Fast Food Restaurant             Restaurant   
15                       Bar             Coffee Shop         Massage Studio   
17                      Café                    Bank    Japanese Restaurant   
18         Electronics Store                 Airport        Warehouse Store   
20             Shopping Mall                    Park                  Hotel   
21                     Trail        Basketball Court                 Bakery   
24          Business Service              Food Truck         Baseball Field   
25        Italian Restaurant             Coffee Shop         Sandwich Place   
26               Coffee Shop          Discount Store         Sandwich Place   
28           Bubble Tea Shop             Pizza Place            Coffee Shop   
29        Athletics & Sports          Discount Store           Liquor Store   
30            Breakfast Spot       Convenience Store          Moving Target   
33                  Pharmacy             Pizza Place            Supermarket   

        5th Most Common Venue      6th Most Common Venue  \
1                Liquor Store              Movie Theater   
2                    Boutique         Miscellaneous Shop   
3  

Cluster 1 shows a wide variety of restaurants and shops.

In [58]:
tr_york_merged.loc[tr_york_merged['Cluster Labels'] == 2, tr_york_merged.columns[[1] + list(range(5, tr_york_merged.shape[1]))]]

Borough  br_n  Cluster Labels 1st Most Common Venue  \
27  North York     2               2          Food Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
27        Baseball Field     Accessories Store             Juice Bar   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
27         Movie Theater     Mobile Phone Shop    Miscellaneous Shop   

        8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
27  Middle Eastern Restaurant    Mexican Restaurant          Metro Station

In [59]:
tr_york_merged.loc[tr_york_merged['Cluster Labels'] == 3, tr_york_merged.columns[[1] + list(range(5, tr_york_merged.shape[1]))]]

Borough  br_n  Cluster Labels 1st Most Common Venue  \
22  North York     2               3           Pizza Place   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22     Accessories Store                  Park         Movie Theater   

   5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
22     Mobile Phone Shop    Miscellaneous Shop  Middle Eastern Restaurant   

   8th Most Common Venue 9th Most Common Venue    10th Most Common Venue  
22    Mexican Restaurant         Metro Station  Mediterranean Restaurant

Cluster 2 and 3 are primarily distinguished through having the only food service and being far detached from the other venues.

## How could this have been improved?

As could be seen, cluster 1 takes too many venues compared to the others, especially 2 and 3. By increasing the range of the radius and the amount of venues, and by selecting different boroughs, the clustering could have become more meaningful.